common/src/main/java/com/wynntils/core/text/PartStyle.java

```diff
@@ -0,0 +1,323 @@
+
+        if (parentStyle == null) {
+            inheritedStyle = style;
+        } else {
+            // This changes properties that are null, as-in, inherting from the previous style.
+            inheritedStyle = style.applyTo(parentStyle);
+
+            // We don't want to inherit these properties.
+            inheritedStyle = inheritedStyle
+                    .withClickEvent(style.getClickEvent())
+                    .withHoverEvent(style.getHoverEvent())
+                    .withInsertion(style.getInsertion())
+                    .withFont(style.getFont());
+        }
+
+        return new PartStyle(
+                owner,
+                inheritedStyle.getColor() == null
+                        ? CustomColor.NONE
+                        : CustomColor.fromInt(inheritedStyle.getColor().getValue()),
+                inheritedStyle.isBold(),
+                inheritedStyle.isItalic(),
+                inheritedStyle.isUnderlined(),
+                inheritedStyle.isStrikethrough(),
+                inheritedStyle.isObfuscated(),
+                inheritedStyle.getClickEvent(),
+                inheritedStyle.getHoverEvent());
+    }
+
+    public String asString(PartStyle previousStyle, StyleType type) {
+        // Rules of converting a Style to a String:
+        // Every style is prefixed with a §.
+        // 0. Every style string is fully qualified, meaning that it contains all the formatting, and reset if needed.
+        // 1. Style color is converted to a color segment.
+        //    A color segment is the prefix and the chatFormatting char.
+        //    If this is a custom color, a hex color code is used.
+        //    Example: §#FF0000 or §1
+        // 2. Formatting is converted the same way as in the Style class.
+        // 3. Click events are wrapped in square brackets, and is represented as an id.
+        //    The parent of this style's owner is responsible for keeping track of click events.
+        //    Example: §[1] -> (1st click event)
+        // 4. Hover events are wrapped in angle brackets, and is represented as an id.
+        //    The parent of this style's owner is responsible for keeping track of hover events.
+        //    Example: §<1> -> (1st hover event)
+
+        if (type == StyleType.NONE) return "";
+
+        StringBuilder styleString = new StringBuilder();
+
+        boolean skipFormatting = false;
+
+        // If the color is the same as the previous style, we can try to construct a difference.
+        // If colors don't match, the inserted color will reset the formatting, thus we need to include all formatting.
+        // If the current color is NONE, we NEED to try to construct a difference,
+        // since there will be no color formatting resetting the formatting afterwards.
+        if (previousStyle != null && (color == CustomColor.NONE || previousStyle.color.equals(color))) {
+            StringBuilder stringBuilder = this.tryConstructDifference(previousStyle);
```

Comment:
It's kind of odd to return a stringBuilder, just to append it to another stringBuilder :)

I recommend you send in the styleString stringbuilder as an argument to tryConstructDifference, and return a boolean saying if you did find a difference or not.

Reply:
I know, but doing that would require changes to the method, which would not turn out so great (since you append, and find return cases later). This means that we would lose the clarity of that method.

Reply:
I see. I missed all `return null` except the first...  Maybe at least make tryConstructDifference return a String..?

Also, in general, for modern Java, StringBuild does not provide any performance gains compared to normal string operations (because of the string indyfy optimizations), so it only makes sense to use it if the code gets clearer from it. In this case I think it does for the "outer" asString() method, at least.

Reply:
In that sense, I think append is much cleaner here than formats += ChatFormatting.RED.

Reply:
StringBuilder is this kind of "dirty" implementation detail that should in general not be shared between methods.

Reply:
Did it.